File: lib/iriq/observation.rb
module Iriq
  # The result of Corpus#observe. Lightweight value object — heavy work
  # (cluster load, explanation, normalization) is deferred until you ask.
  class Observation
    attr_reader :identifier

    def initialize(corpus:, identifier:, cluster_key:)
      @corpus      = corpus
      @identifier  = identifier
      @cluster_key = cluster_key
    end

    # The cluster as it stands when first read; observing doesn't load it.
    def cluster
      @cluster ||= @corpus.storage.cluster_for(@cluster_key)
    end

    def fingerprint
      @fingerprint ||= Normalizer.normalize_identifier(@identifier)
    end

    def explanation
      @explanation ||= @corpus.explain(@identifier)
    end

    def normalize
      @corpus.normalize(@identifier)
    end
  end
end

File: observation.go
package iriq

// Observation is the result of Corpus.Observe.
type Observation struct {
	corpus      *Corpus
	Identifier  *Identifier
	Cluster     *Cluster
	fingerprint string
	explanation []CorpusEntry
}

// Fingerprint returns the deterministic (non-corpus-aware) normalized form.
func (o *Observation) Fingerprint() string {
	if o.fingerprint == "" {
		o.fingerprint = NormalizeIdentifier(o.Identifier, DefaultClassifier, true)
	}
	return o.fingerprint
}

// Explanation returns the corpus-informed explanation. Memoized.
func (o *Observation) Explanation() []CorpusEntry {
	if o.explanation == nil {
		o.explanation = o.corpus.Explain(o.Identifier)
	}
	return o.explanation
}

func (o *Observation) Normalize() string {
	return o.corpus.NormalizeIdentifier(o.Identifier)
}
